### ISSUE TITLE
`ncol` and `nrow` seems to be in a reverse order

### ISSUE THREAD

**Issue by Nowosad** [r-spatial/stars#613]
Hi @edzer, I was chasing a bug in a package that uses **stars**, when I found that **stars** return different values for `ncol()` and `nrow()` than I would expect (and also different from the **terra** output). Is it a bug or do I miss something here?

``` r
# devtools::install_github("r-spatial/stars")
library(stars)
#> Loading required package: abind
#> Loading required package: sf
#> Linking to GEOS 3.11.0, GDAL 3.5.2, PROJ 9.0.1; sf_use_s2() is TRUE
library(terra)
#> terra 1.7.3
x = read_stars(system.file("tif/L7_ETMs.tif", package = "stars"))
r = rast(system.file("tif/L7_ETMs.tif", package = "stars"))
plot(x, axes = TRUE)
```

![](https://i.imgur.com/ScIoHri.png)

``` r

# stars
ncol(x)
#>   y 
#> 352
nrow(x)
#>   x 
#> 349

# terra
ncol(r)
#> [1] 349
nrow(r)
#> [1] 352
```

**Comment by edzer** [comment]
I can see why you would like this to be different, but `base::nrow(x)` is defined as `dim(x)[1L]`, so so be it. It may be more useful to think of the number elements in `x` direction and use `dim(x)["x"]`, of course there may be datasets where that dimension is called `lon` or `longitude`. Also, 
```r
> aperm(x, c(2,1,3)) |> nrow()
  y 
352 
```
The logic behind this is that matrices / arrays are row-major in R, meaning rows cycle fastest; in imagery storage typically the columns cycle fastest (although this may vary, e.g. in NetCDF). Following the storage form rather than adopting a row/col reversal logic (as in `sp`) removed a major PITA from a coding perspective, IMO.

TL;DR: use `dim`, forget about `ncol` / `nrow`.

**Comment by Nowosad** [comment]
Thank you for the explanation, @edzer -- it helps a lot.
